src/metabase/upload.clj

```diff
@@ -2,37 +2,68 @@
   (:require
    [clojure.data.csv :as csv]
    [clojure.java.io :as io]
+   [clojure.set :as set]
    [clojure.string :as str]
    [flatland.ordered.map :as ordered-map]
+   [flatland.ordered.set :as ordered-set]
    [java-time :as t]
    [metabase.driver :as driver]
    [metabase.search.util :as search-util]
-   [metabase.util :as u]
-   [metabase.util.i18n :refer [trs]]))
+   [metabase.util :as u]))
 
 (set! *warn-on-reflection* true)
 
 ;;;; +------------------+
 ;;;; | Schema detection |
 ;;;; +------------------+
 
-;;           text
-;;            |
-;;            |
-;;       varchar_255
-;;            |
-;;            |
-;;          float
-;;            |
-;;            |
-;;           int
-;;            |
-;;            |
-;;         boolean
-(derive ::boolean     ::int)
-(derive ::int         ::float)
-(derive ::float       ::varchar_255)
-(derive ::varchar_255 ::text)
+;;              text
+;;               |
+;;               |
+;;          varchar_255
+;;              / \
+;;             /   \
+;;            /     \
+;;         float   datetime
+;;           |       |
+;;           |       |
+;;          int    date
+;;           |
+;;           |
+;;        boolean
+
+(def type->parent
+  ;; listed in depth-first order
+  {::varchar_255 ::text
+   ::float       ::varchar_255
+   ::int         ::float
+   ::boolean     ::int
+   ::datetime    ::varchar_255
+   ::date        ::datetime})
```

Comment:
I reimplemented the hierarchy using a map, because it enforces that each type has exactly one parent. It meant I could implement `lowest-common-ancestor` more efficiently too. Instead of using `parents` I can use `type->parent`.